### PR TITLE
Fix/17966 and 17967

### PR DESCRIPTION
**Pull request by tienifr**
### Details

Here is the flow when we open 2 tabs and logout on first tab:
- call logout API on first tab
- clear onyx data on first tab
- sync data on second tab
- that makes the components on second tab re-render (ReportScreen)
- call API `OpenReport`, but at that time the session is expired

We should remove Session first before signing out. We also should remove Credential first to prevent second tab back to magicCode page (just a moment).

Cleaning up reconnectApp callback in  `AuthScreen` to prevent re-call `ReconnectApp` API in second tab

### Fixed Issues

$ https://github.com/Expensify/App/issues/15321
$ https://github.com/Expensify/App/issues/17967
$ https://github.com/Expensify/App/issues/17966

### Tests

- Verify bug 17966
1. Open the app in the browser
2. Open another tab in the same browser
3. Log out from one of the tabs
4. Verify that in other tap the app is logged out.
5. Log in with the same user
6. Tap "Didn't receive the magic code"
7. Verify infinite spinner does not appear in login page after tap didn't receive a magic code
- Verify bug 17967
1. Open the app in the browser
2. Open another tab in the same browser
3. Disable the internet
4. Log out from one of the tabs
5. Verify that in other tab the app is not crashed and logged out to the login page.
6. Verify Skeleton does not appear for a while before log out automatically on a second tab
7. Login in with the same user
8. Verify everything works as expect

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

- Verify bug 17966
1. Open the app in the browser
2. Open another tab in the same browser
3. Log out from one of the tabs
4. Verify that in other tap the app is logged out.
5. Log in with the same user
6. Tap "Didn't receive the magic code"
7. Verify infinite spinner does not appear in login page after tap didn't receive a magic code
- Verify bug 17967
1. Open the app in the browser
2. Open another tab in the same browser
3. Disable the internet
4. Log out from one of the tabs
5. Verify that in other tab the app is not crashed and logged out to the login page.
6. Verify Skeleton does not appear for a while before log out automatically on a second tab
7. Login in with the same user
8. Verify everything works as expect
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/113963320/234586183-3d22a7e8-2d1d-4b8f-8dbd-f679ece12194.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/113963320/234599207-f170e66c-39da-4598-8436-427072a75f57.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/234592978-7dc07b97-573e-4e2d-9096-efd58f991ec3.mp4

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by tienifr** [comment]
@sobitneupane @roryabraham I've created the PR to fix some regressions. Can you please help review? Thanks

**Review by roryabraham — DISMISSED** [review]
LGTM 👍🏼

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/25876548/234810663-4956453c-23bd-403a-8aa3-202051881793.mov


https://user-images.githubusercontent.com/25876548/234810811-f7701d0e-d2f9-47c3-b6c0-5f41946d00a1.mov



</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/25876548/234809490-5c7e46aa-5dd5-4c12-a651-35ca0c2e8c72.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>


https://user-images.githubusercontent.com/25876548/234813873-0fc0ab2b-2e27-446a-9e81-e36f0146a394.mp4


https://user-images.githubusercontent.com/25876548/234813970-57fa0d1c-f487-43e3-81b4-3756cacce6d6.mp4


</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/25876548/234809202-f059d346-5aa9-4331-a8dc-80bf5fcaed3d.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/25876548/234809408-e0fcb517-1cc8-415c-a0b4-802beff09936.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/25876548/234812423-a167006f-8f09-415c-81ab-633dfcfdfe8b.mov

</details>

**Comment by sobitneupane** [comment]
I can see some issues in mWeb-Safari. But these issues are not always reproducible. I was able to reproduce them once or twice.

- When I logout, `Magic code` page appears for a second or so.

https://user-images.githubusercontent.com/25876548/234817408-6276751b-6561-4394-8319-2388e6fa26ef.MP4


- When I logout, error (Incorrect Magic code) shown on login screen for few seconds.

- Unread Indicator reappears.

**Comment by tienifr** [comment]
@sobitneupane I find the problem. I reseted both credential and session, that can make session is updated on second devices first -> redirect to signin page(at that time we still have stale credential) -> magic code screen is shown for a second.

That why we I add a new field (email) in session when the API `BeginSignIn` calls successfully. And redirect to magic code screen if we have the email in session, so we can make sure when we clear session, the email screen is shown immediately.

Here is my result:

https://user-images.githubusercontent.com/113963320/234844782-7b6b8fd0-6fc8-41ad-b0ac-1fa8bcf729dc.mp4

**Comment by sobitneupane** [comment]
@tienifr Thanks for the update

Following issue is still reproducible.

> Unread Indicator reappears.


https://user-images.githubusercontent.com/25876548/235141636-e1bf3c90-a846-4f20-9f72-56a9e2785578.mp4

**Comment by tienifr** [comment]
@sobitneupane I want to explain the issue here:
1. User opens 3 tabs
2. Login in tab 1
3. API `OpenApp` in tab 1 is called first
4. Update onyx data from `OpenApp` API
5. Sync onyx data to other tabs 
6. Logout in tab 1
7. API `OpenApp` in tab 2 is called successfully
8. Update onyx data from `OpenApp` API

That why we see the unread Indicator reappears.
Please check the video below:

https://user-images.githubusercontent.com/113963320/235208842-35eaf141-5867-4f14-ace0-e7a606d53a61.mp4

I also found another problem when users sign in with magic code:
1. User opens 3 tabs
2. Login in tab 1
3. API `SigninUser` is pushed to request queue
4. Sync request queue to other tabs
5. Request queue in tab 2, 3 has `SigninUser`
6. Call `SigninUser` API successfully in tab 1
7. Remove `SigninUser` from queue (tab 1)
8. Sync request queue to other tabs
9. Switch to tab 2
10. Flush request queue in tab 2 (at that time the `SigninUser` is still in queue)
11. `SigninUser` is called 1 more time

**Comment by tienifr** [comment]
Do you think it is the problem of our architecture when sync onyx data between multiple tabs? If there is, I think it's out of PR's scope? Thanks

**Comment by sobitneupane** [comment]
@tienifr This shows that our latest changes in Onyx should have probably introduced [this issue](https://github.com/Expensify/App/pull/18027#issuecomment-1527462390) while trying to solve other. But it is a small issue and might not occur in real use case. @roryabraham What's your thought on this [issue](https://github.com/Expensify/App/pull/18027#issuecomment-1527462390)?

**Comment by sobitneupane** [comment]
@tienifr This shows that our latest changes in onyx has created [this issue](https://github.com/Expensify/App/pull/18027#issuecomment-1527462390) while trying to solve other. Let's get this fixed, else it will create some unwanted issues if onyx doesn't get cleared properly.
Screenshot 2023-05-03 at 12 34 38

**Comment by tienifr** [comment]
@sobitneupane I've updated my PR. The problem here is because of sync process between tabs (After clearing IndexDB in tab 1, API results come from tab 2 can update IndexDB)

1. On Onyx repo:

abort pending Onyx.setItem when we clear data
https://github.com/Expensify/react-native-onyx/pull/248/files

2. On App repo:

As we call `OpenApp` API in both tabs so I think we also should call Onyx.clear() in all tabs. When users sign-out, I update the session data to {isLogginOut:true}, this value will be synced to other tabs then in `SignInRedirect.js` we will execute Onyx.clear()

Here's the result. Please tell me what your though is. Thanks

https://user-images.githubusercontent.com/113963320/236821202-b2ac0e03-c0c4-4b94-a775-269e22a7618a.mp4

**Comment by tienifr** [comment]
> Those issues were not existent before

I agree because we can't test it before, the second tab will be crashed when logging out in first tab. 
After spending time to understand onyx and the root cause, I can't find the solution to solve the issue in onyx. (The root cause is response of OpenApp API in tab 2 updates IndexDB after logging out in tab 1)
Can anyone help me this edge case cc @sobitneupane @roryabraham? Thanks

**Comment by sobitneupane** [comment]
> I agree because we can't test it before, the second tab will be crashed when logging out in first tab.

We can and I did test it with [these changes](https://github.com/Expensify/App/pull/15556/files).

**Comment by roryabraham** [comment]
@tienifr One thing that you can do that I think will _help_ with the problem is by updating [this code](https://github.com/Expensify/App/blob/ab5862a4fcc68f4712be325d6c68f0640f0521f8/src/libs/actions/User.js#L580-L585) in `User.subscribeToUserEvents`:

```diff
diff --git a/src/libs/actions/User.js b/src/libs/actions/User.js
index fbe3f691cd..3ee87809c1 100644
--- a/src/libs/actions/User.js
+++ b/src/libs/actions/User.js
@@ -21,6 +21,7 @@ import * as ReportActionsUtils from '../ReportActionsUtils';
 import DateUtils from '../DateUtils';
 import * as Session from './Session';
 import * as PersonalDetails from './PersonalDetails';
+import * as ActiveClientManager from '../ActiveClientManager';
 
 let currentUserAccountID = '';
 let currentEmail = '';
@@ -579,6 +580,10 @@ function subscribeToUserEvents() {
     // Receive any relevant Onyx updates from the server
     PusherUtils.subscribeToPrivateUserChannelEvent(Pusher.TYPE.ONYX_API_UPDATE, currentUserAccountID, (pushJSON) => {
         SequentialQueue.getCurrentRequest().then(() => {
+            if (!ActiveClientManager.isClientTheLeader()) {
+                Log.client('Skipping Onyx update event because client is not the leader');
+                return;
+            }
             Onyx.update(pushJSON);
             triggerNotifications(pushJSON);
         });
```

That should help with some edge cases where an Onyx update is received by a background tab when the leader tab is already logged out. It doesn't seem to fully resolve all the issues though. 

---

Otherwise, I'm also struggling to find the best way to solve these problems. Some things I've tried are:

- **Theory:** the cause of this problem might be that other tabs have sent `OpenApp` requests that are still in flight when we sign out on the first tab, so clearing those out would prevent their callbacks from inserting data into Onyx.
- **Solution:** Update [`cleanupSession`](https://github.com/Expensify/App/blob/615ebd98c4b32affe2875712187d202a7dc64ab0/src/libs/actions/Session/index.js#L556-L560) so that it also clears the network request queues and any pending HTTP requests

- **theory:** `CleanupSession` is called when `AuthScreens` unmounts. That was taking too long or not happening for background tabs, so pending requests are not cancelled and their data is inserted into Onyx
- **solution:** Use the [`BroadcastChannel` API](https://developer.mozilla.org/en-US/docs/Web/API/BroadcastChannel) to quickly call `CleanupSession` in other tabs when `SignOut` is call on any tab

- **theory:** The `LogOut` API command is cancelled before it finishes, and this was creating some problems for other tabs
- **solution:** Use `makeRequestWithSideEffects` to force the `LogOut` API command to finish before clearing all the session data and redirecting (bad practice, pulling at straws at this point). Also tried clearing the main and sequential request queues before calling `LogOut`.

- **theory:** The problem is happening at the Onyx layer – `Onyx.update` is called in one tab at the same time as `Onyx.clear` in other tabs, and that causes a race condition where sometimes `Onyx.clear` finishes first and another times other Onyx updates finish first.
- **solution:** TBD, but let's discuss more

---

> abort pending Onyx.setItem calls when we clear data

I briefly looked at [this PR](https://github.com/Expensify/react-native-onyx) and seems like maybe a step in the right direction, but I don't know if it will solve the problem across multiple tabs. I don't think the `SyncQueue` is shared between tabs.

> As we call OpenApp API in both tabs so I think we also should call Onyx.clear() in all tabs. When users sign-out, I update the session data to {isLogginOut:true}, this value will be synced to other tabs then in SignInRedirect.js we will execute Onyx.clear()

I see this is your proposed solution for syncing between tabs.

I kind of prefer the idea of fixing this synchronization problem in the Onyx layer itself, like @sobitneupane suggestion. That way, we don't have to manage it so carefully in the app space. Solving the problem in Onyx also would mean that we could get a more generic solution that works in all cases and in different apps.

One idea to synchronize Onyx changes across tabs is to create a single shared Onyx `SyncQueue`, that's shared between tabs using the `BroadcastChannel` API. That way, if `Onyx.clear` is called on any of the tabs, we could abort queued set/merge operations in all the other tabs, clear the data from disk, and update the subscribers across all tabs.

Overall I think we should talk about this more in slack, but it's very late for me here and I need a clearer head to gather my thoughts.

**Comment by tienifr** [comment]
@roryabraham What is our next step? Can you please open Slack thread

**Comment by tienifr** [comment]
@roryabraham any updates here?

**Comment by roryabraham** [comment]
Sorry for the radio silence @tienifr, this is a bit of a complicated problem, but hard to prioritize since it's such an edge case. We've been discussing it in [this slack thread](https://expensify.slack.com/archives/C03U7DCU4/p1686588535455649), but that's an internal thread.

As stated [here](https://github.com/Expensify/App/issues/15321#issuecomment-1585113844), I think we should pay you out for this PR, then take a step back and try to evaluate the problem more holistically.

I could really use some help nailing down a solid problem statement for the root cause of the issues we were seeing here, because I'm not really confident in my understanding there

**Comment by roryabraham** [comment]
@tienifr I'm going to close this out for now. We'll continue in the issue
